### PR TITLE
Add install_certificates command

### PR DESCRIPTION
**Pull request by thomasw**
This adds an `install_certificates` command to yodeploy which allows it to download certificates from a remote certificate service and store them locally for use by configured applications. This command additionally provides a --clean option allowing users to purge previously downloaded certificates.

---

This supersedes #228. Most of the history from @stormf's branch has been squashed and the commit messages have been slightly altered.

For reference, the original history is still available here: https://github.com/yola/yodeploy/tree/thomas/certificates

There are still some outstanding questions about whether or not this functionality should be included in yodeploy and about how exactly it should be included (https://github.com/yola/production/issues/3810#issuecomment-535735916). The history might need to be rewritten again depending on the conclusions reached in that discussion.

### CODE REVIEW

File: yodeploy/cmds/install_certificates.py [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+        for f in os.listdir(path):
+            full_path = os.path.join(path, f)
+
+            if os.path.isfile(full_path):
+                yield full_path
+
+
+def remove_all_files(files):
+    for file in files:
+        print('removing {}'.format(file))
```
Comment (snitch) [inline yodeploy/cmds/install_certificates.py]: T001 print found.

File: yodeploy/cmds/install_certificates.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+    clean_message = '{} and {} are clean.'.format(certs_dir, private_dir)
+
+    # Take care here not to generate this list before and again after the
+    # confirmation prompt. If one were to do so and a file was written to a
+    # target directory before the remove operation executed, we would end up
+    # removing something the user hadn't approved
+    to_remove = list(get_files(certs_dir, private_dir))
+
+    if not to_remove:
+        print(clean_message)
```
Comment (snitch) [inline yodeploy/cmds/install_certificates.py]: T001 print found.

File: yodeploy/cmds/install_certificates.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+        return
+
+    if confirm:
+        for f in to_remove:
+            print(f)
```
Comment (snitch) [inline yodeploy/cmds/install_certificates.py]: T001 print found.

File: yodeploy/cmds/install_certificates.py [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+
+        if response == 'y':
+            break
+
+        if response == 'n':
+            return
+
+    remove_all_files(to_remove)
+
+    print(clean_message)
```
Comment (snitch) [inline yodeploy/cmds/install_certificates.py]: T001 print found.

File: yodeploy/cmds/install_certificates.py [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+
+    deploy_config = config.find_deploy_config()
+    settings = config.load_settings(deploy_config)
+    cert_service = CertificateService(
+        settings.envhub.url,
+        auth=(settings.envhub.user, settings.envhub.password))
+
+    if options.list:
+        for env in cert_service.get_environments():
+            print('{name} -- {developer}'.format(**env))
```
Comment (snitch) [inline yodeploy/cmds/install_certificates.py]: T001 print found.

File: yodeploy/cmds/install_certificates.py [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+        pk_local_path = os.path.join(ssl_dir, 'private')
+
+        cert_local_path = os.path.join(cert_local_path, cert_filename)
+        pk_local_path = os.path.join(pk_local_path, key_filename)
+
+        ensure_required_ssl_subdirs_exist(ssl_dir)
+
+        with open(cert_local_path, 'w') as fn:
+            fn.write(cert)
+            print('certificate saved as {}'.format(cert_local_path))
```
Comment (snitch) [inline yodeploy/cmds/install_certificates.py]: T001 print found.

File: yodeploy/cmds/install_certificates.py [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+
+        with open(pk_local_path, 'w') as fn:
+            fn.write(key)
+            print('private key saved as {}'.format(pk_local_path))
```
Comment (snitch) [inline yodeploy/cmds/install_certificates.py]: T001 print found.

File: yodeploy/cmds/install_certificates.py [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+from demands import HTTPServiceClient
+
+sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', '..'))
+
+from yodeploy import config  # noqa
+
+# overrwrite the unsafe python 2 input method with raw_input, which does not
+# eval user input. In Python 3, raw_input has been renamed to input. This
+# assignment will raise a NameError, doing nothing.
+try:
```
Comment (thomasw) [inline yodeploy/cmds/install_certificates.py]: We're doing this in at least one other place.

six is currently a transitive dependency. I don't mind making that explicit and using `six.moves.input` here

File: CHANGELOG.md [hunk 9]
```diff
@@ -1,6 +1,12 @@
 Change Log
 ==========
 
+0.11.0
+------
+
+* Add install_certificate command, enabling consumers to download certificates
+  for environments from a remote certificate store.
```
Comment (stefanor) [inline CHANGELOG.md]: This isn't part of the library, I don't think it needs a new version (or mention in CHANGELOG)

File: yodeploy/cmds/install_certificates.py [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+    input = raw_input
+except NameError:
+    pass
+
+
+class CertificateService(HTTPServiceClient):
+    def _get_filename_from_headers(self, headers, fallback):
```
Comment (stefanor) [inline yodeploy/cmds/install_certificates.py]: Why even bother with this, if we have a reasonable filename?

File: yodeploy/cmds/install_certificates.py [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+
+def main():
+    parser = argparse.ArgumentParser(
+        description='Downloads ssl certificates from envhub and installs them '
+                    'locally for development purposes.'
+    )
+    parser.add_argument('-l', '--list', action='store_true',
+                        help='List instance_ids of running envs')
+    parser.add_argument('-i', '--install',
+                        help='Install ssl certificate for given instance name')
```
Comment (stefanor) [inline yodeploy/cmds/install_certificates.py]: `metavar='INSTANCE'` ?

**Pull request closed without merging**